File: frontend/src/types/fsm.ts
// FSM Type Definitions

export type FSMType = 'moore' | 'mealy';
export type Visibility = 'private' | 'public' | 'unlisted';

export interface State {
  id: string;
  name: string;
  output?: string;
  position?: { x: number; y: number };
  is_dummy?: boolean;
}

export interface Transition {
  id?: string;
  from_state: string;
  to_state: string;
  input?: string;
  output?: string;
  label?: string;
}

export interface FSMDefinition {
  states?: State[];
  transitions?: Transition[];
  outputs?: Record<string, string>;
}

export interface FSM {
  id: string;
  name: string;
  description?: string;
  fsm_type: FSMType;
  states: string[];
  initial_state: string;
  transitions: Transition[];
  outputs?: Record<string, string>;
  encoding?: Record<string, string>;
  definition: FSMDefinition;
  state_count: number;
  transition_count: number;
  bit_width?: number;
  tags?: string[];
  visibility: Visibility;
  created_at?: string;
  updated_at?: string;
  created_by?: string;
  view_count?: number;
  fork_count?: number;
  is_optimized?: boolean;
  dummy_state_count?: number;
  optimization_algorithm?: string;
  avg_hamming_distance?: number;
  encoding_type?: string;
}

export interface FSMCreate {
  name: string;
  description?: string;
  fsm_type: FSMType;
  states: string[];
  initial_state: string;
  transitions: Transition[];
  outputs?: Record<string, string>;
  tags?: string[];
  visibility?: Visibility;
}

export interface FSMUpdate extends Partial<FSMCreate> {
  id?: string;
}

export interface DummyState {
  id: string;
  encoding: string;
  output?: string;
  is_dummy: boolean;
  inserted_for: string;
}

export interface OptimizedFSM extends FSM {
  dummy_states: DummyState[];
  algorithm_used: string;
  // Per-algorithm metrics. Common keys: avg_hamming_before/_after,
  // max_hamming, improvement_percentage. Kept loose because each
  // algorithm contributes its own keys.
  metrics: Record<string, number | string | boolean>;
  optimization_time_ms: number;
}

export interface AlgorithmResult {
  id?: string;
  // Set when the algorithm produced a derived FSM (greedy/BFS may add
  // dummy states; SA/GA produce a re-encoded FSM). Lets the lab report
  // re-fetch the optimized FSM after revisiting the page.
  optimized_fsm_id?: string | null;
  algorithm: string;
  dummy_states_added: number;
  total_states_final: number;
  avg_hamming_before: number;
  avg_hamming_after: number;
  improvement_percentage: number;
  execution_time_ms: number;
  encoding_map: Record<string, string>;
  executed_at?: string;
}

export interface OptimizationRequest {
  algorithm: 'greedy' | 'bfs_optimal' | 'global_sa' | 'global_ga';
  options?: {
    timeout_ms?: number;
    iterations?: number;
    temperature?: number;
    cooling_rate?: number;
  };
  async_mode?: boolean;
}

export interface OptimizationMetrics {
  avg_hamming_before: number;
  avg_hamming_after: number;
  max_hamming_before: number;
  max_hamming_after: number;
}

export interface OptimizationResponse {
  optimized_fsm_id: string;
  algorithm: string;
  execution_time_ms: number;
  dummy_states_added: number;
  total_states: number;
  improvement_percentage: number;
  metrics: OptimizationMetrics;
  encoding_map?: Record<string, string>;
}

export interface ExportFormat {
  format: 'verilog' | 'vhdl' | 'json' | 'csv' | 'testbench';
  options?: {
    module_name?: string;
    include_comments?: boolean;
    style?: 'standard' | 'compact' | 'verbose';
  };
}

export interface ExportRequest extends ExportFormat {}

export interface ExportResponse {
  format: string;
  content: string;
  file_name: string;
}

File: backend/app/schemas/fsm.py
"""
Pydantic schemas for FSM API requests/responses
"""

import re
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator


class TransitionBase(BaseModel):
    """Base transition schema"""

    from_state: str
    to_state: str
    input: str | None = None
    # HDL output signal value — restricted to binary/don't-care characters so
    # that user-supplied values cannot be injected verbatim into Verilog/VHDL.
    output: str | None = Field(default=None, pattern=r"^[01xXzZ-]*$")
    label: str | None = None


class FSMCreate(BaseModel):
    """Schema for creating a new FSM"""

    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    fsm_type: str = Field(..., pattern="^(moore|mealy)$")
    states: list[str] = Field(..., min_length=1)
    initial_state: str
    transitions: list[dict[str, Any]]
    # Moore outputs keyed by state name — values are constrained to binary/don't-care
    # characters to prevent HDL injection when values are written into generated code.
    outputs: dict[str, str] | None = Field(default=None)

    @field_validator("outputs")
    @classmethod
    def validate_outputs(cls, v: dict[str, str] | None) -> dict[str, str] | None:
        """Ensure output values only contain binary/don't-care characters."""
        if v is None:
            return v
        _hdl_pattern = re.compile(r"^[01xXzZ-]*$")
        for state, val in v.items():
            if not _hdl_pattern.match(val):
                raise ValueError(
                    f"Output value for state '{state}' contains invalid characters. "
                    "Only binary digits and don't-cares (0, 1, x, X, z, Z, -) are allowed."
                )
        return v

    category_id: UUID | None = None
    tags: list[str] | None = []
    visibility: str = Field(default="private", pattern="^(private|public|unlisted)$")

    @field_validator("initial_state")
    @classmethod
    def validate_initial_state(cls, v: str, info: Any) -> str:
        """Validate initial state is in states list"""
        if "states" in info.data and v not in info.data["states"]:
            raise ValueError(f"Initial state '{v}' not in states list")
        return v


class FSMResponse(BaseModel):
    """Schema for FSM response"""

    id: UUID
    name: str
    description: str | None
    fsm_type: str
    state_count: int
    transition_count: int
    initial_state: str
    bit_width: int
    is_optimized: bool
    dummy_state_count: int
    view_count: int
    states: list[str] = []
    # Expose transitions so the editor can round-trip the full FSM (without
    # this, a saved FSM reloads with no transitions).
    transitions: list[dict[str, Any]] = []
    created_at: datetime
    updated_at: datetime | None

    class Config:
        from_attributes = True


class FSMUpdate(BaseModel):
    """Schema for updating an existing FSM"""

    name: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = None
    tags: list[str] | None = None
    visibility: str | None = Field(None, pattern="^(private|public|unlisted)$")
    # Definition edits so the editor's Save persists the actual FSM, not just
    # metadata. All optional — omitted fields are left untouched.
    states: list[str] | None = Field(None, min_length=1)
    initial_state: str | None = None
    transitions: list[dict[str, Any]] | None = None
    outputs: dict[str, str] | None = None


class FSMFork(BaseModel):
    """Schema for forking an existing FSM"""

    name: str = Field(..., min_length=1, max_length=255)


class OptimizationRequest(BaseModel):
    """Request schema for FSM optimization"""

    model_config = ConfigDict(populate_by_name=True)

    algorithm: str = Field(..., pattern="^(greedy|bfs_optimal|global_sa|global_ga)$")
    options: dict[str, Any] | None = {}
    async_mode: bool = Field(default=False, alias="async")


class OptimizationMetrics(BaseModel):
    """Hamming distance metrics for optimization results"""

    avg_hamming_before: float
    avg_hamming_after: float
    max_hamming_before: int
    max_hamming_after: int


class OptimizationResponse(BaseModel):
    """Response schema for optimization"""

    optimized_fsm_id: UUID
    algorithm: str
    execution_time_ms: int
    dummy_states_added: int
    total_states: int
    improvement_percentage: float
    metrics: OptimizationMetrics
    encoding_map: dict[str, str] | None = None
